[Validator] Handle object properties in Unique validator

plfort opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #48072
| License       | MIT

The `fields` option of `Unique` only read array keys, so a collection of objects could not be checked for uniqueness on a property. Objects are now read as well.

```php
#[Assert\Unique(fields: ['sku'])]
private array $items;   // [new Item('a'), new Item('a')] is now a violation
```

Public properties are read directly. Anything else goes through PropertyAccess, so getters work and so do property paths:

```php
#[Assert\Unique(fields: ['fieldA[fieldB]'])]
```

Fields that cannot be read stay optional, exactly as they already are for arrays: the element is compared on the fields that could be read. Non-array, non-object elements such as `''`, `1` and `null` keep being skipped when `fields` is set.

`UniqueValidator` takes an optional `PropertyAccessorInterface`, and builds one on demand otherwise, the same way `AbstractComparisonValidator` and `RangeValidator` already do. `symfony/property-access` stays a dev dependency; the validator throws a clear `LogicException` naming the package if a property path is used without it.

Documentation should cover: `fields` now accepting object properties, that public properties are read directly and everything else through PropertyAccess, that property paths are supported, and that unreadable fields are ignored rather than raising.

plfort committed:
Handle object field in UniqueValidator

plfort committed:
Change CHANGELOG entry order

Fix after rebase

Update CHANGELOG.md

plfort committed:
CS fix

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/UniqueValidator.php:
```diff
@@ -21,6 +26,13 @@
  */
 class UniqueValidator extends ConstraintValidator
 {
+    private ?PropertyAccessorInterface $propertyAccessor;
+
+    public function __construct(PropertyAccessorInterface $propertyAccessor = null)
+    {
+        $this->propertyAccessor = $propertyAccessor;
+    }
```
As the minimal requirement for 6.3 is PHP 8.1, I guess this could be using constructor property promotion 👌

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/UniqueValidator.php:
```diff
@@ -69,18 +81,46 @@ private function getNormalizer(Unique $unique): callable
+    {
+        if (null === $this->propertyAccessor) {
+            if (!class_exists(PropertyAccess::class)) {
+                throw new LogicException('Unable to use property path as the Symfony PropertyAccess component is not installed.');
```
Maybe?

```suggestion
                throw new LogicException('Property path requires symfony/property-access package to be installed.');
```

plfort committed:
Use constructor property promotion

plfort committed:
Update src/Symfony/Component/Validator/Constraints/UniqueValidator.php

Co-authored-by: Alexandre Daubois <2144837+alexandre-daubois@users.noreply.github.com>

artyuum commented on src/Symfony/Component/Validator/Constraints/UniqueValidator.php:
```diff
@@ -69,18 +81,46 @@ private function getNormalizer(Unique $unique): callable
+    {
+        if (null === $this->propertyAccessor) {
+            if (!class_exists(PropertyAccess::class)) {
+                throw new LogicException('Property path requires symfony/property-access package to be installed.');
```
It would be great for DX if we could write the message as follow:
```suggestion
                throw new LogicException('Property path requires symfony/property-access package to be installed. Try running "composer require symfony/property-access".');
```
Additionally, we would keep the consistency with the other messages as seen here: https://github.com/symfony/symfony/search?q=%22composer+require%22

plfort committed:
Update src/Symfony/Component/Validator/Constraints/UniqueValidator.php

Co-authored-by: Artyum <didifsx@hotmail.fr>

plfort committed:
CS fix

Seb33300 commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -3,7 +3,7 @@ CHANGELOG
 6.3
 ---
-
+ * Add support for reading objects properties with `Unique` constraint `fields` option
```
You should not remove the newline between `---` and he first element of the list.
And usually, we add new element at the end of the list.

plfort replied:
According to https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry 
`New entries must be added on top of the list.`

Seb33300 replied:
Wow right, my bad, so I did it wrong all that time 😅

plfort committed:
Add newline in CHANGELOG.md

plfort committed:
Add validate return type

plfort commented:
Rebased on 7.2
Does everything look OK ? @nicolas-grekas

alexandre-daubois commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -61,6 +61,7 @@ CHANGELOG
 6.3
 ---
 
+ * Add support for reading objects properties with `Unique` constraint `fields` option
```
Can you move this line to the 7.3 section please? Also, it should be appended to the list, not prepended 🙂

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/UniqueValidator.php:
```diff
@@ -21,6 +26,11 @@
 class UniqueValidator extends ConstraintValidator
 {
+    public function __construct(
+        private ?PropertyAccessorInterface $propertyAccessor = null
```
```suggestion
        private ?PropertyAccessorInterface $propertyAccessor = null,
```

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/UniqueValidator.php:
```diff
@@ -64,18 +74,46 @@ private function getNormalizer(Unique $unique): callable
+            $this->propertyAccessor = PropertyAccess::createPropertyAccessor();
+        }
+
+        return $this->propertyAccessor;
```
```suggestion
        if (!class_exists(PropertyAccess::class)) {
            throw new LogicException('Property path requires symfony/property-access package to be installed. Try running "composer require symfony/property-access".');
        }

        return $this->propertyAccessor ??= PropertyAccess::createPropertyAccessor();
```

plfort committed:
Apply suggestions from code review

Co-authored-by: Alexandre Daubois <2144837+alexandre-daubois@users.noreply.github.com>

plfort committed:
Misplaced changelog entry

plfort committed:
[Validator] Handle object properties in the Unique constraint

The fields option only read array keys, so a collection of objects could not
be checked for uniqueness on a property. Objects are now read too, directly
for public properties and through PropertyAccess otherwise, which also brings
property paths such as "fieldA[fieldB]". Fields that cannot be read stay
optional, as they already are for arrays.
